airflow/executors/mesos_executor.py

```diff
@@ -0,0 +1,422 @@
+            raise AirflowException("mesos.docker_volume_logs_name not provided for mesos executor")
+
+        if not conf.get('mesos', 'DOCKER_VOLUME_LOGS_CONTAINER_PATH'):
+            self.log.error("Expecting docker volume logs container path for mesos executor")
+            raise AirflowException("mesos.docker_volume_logs_container_path not provided for mesos executor")
+
+        if not conf.get('mesos', 'DOCKER_SOCK'):
+            self.log.error("Expecting docker sock path for mesos executor")
+            raise AirflowException("mesos.docker_sock not provided for mesos executor")
+
+        self.mesos_slave_docker_image = conf.get('mesos', 'DOCKER_IMAGE_SLAVE').replace('"', '')
+        self.mesos_docker_volume_driver = conf.get('mesos', 'DOCKER_VOLUME_DRIVER').replace('"', '')
+        self.mesos_docker_volume_dag_name = conf.get('mesos', 'DOCKER_VOLUME_DAG_NAME').replace('"', '')
+        self.mesos_docker_volume_dag_container_path = conf.get(
+            'mesos', 'DOCKER_VOLUME_DAG_CONTAINER_PATH'
+        ).replace('"', '')
+        self.mesos_docker_volume_logs_name = conf.get('mesos', 'DOCKER_VOLUME_LOGS_NAME').replace('"', '')
+        self.mesos_docker_volume_logs_container_path = conf.get(
+            'mesos', 'DOCKER_VOLUME_LOGS_CONTAINER_PATH'
+        ).replace('"', '')
+        self.mesos_docker_sock = conf.get('mesos', 'DOCKER_SOCK')
+        self.core_sql_alchemy_conn = conf.get('core', 'SQL_ALCHEMY_CONN')
+        self.core_fernet_key = conf.get('core', 'FERNET_KEY')
+
+    def resource_offers(self, offers):
+        """If we got a offer, run a queued task"""
+        self.log.debug('MESOS OFFER')
+        for i, offer in enumerate(offers):
+            if i == 0:
+                self.run_job(offer)
+            offer.decline()
+            i += 1
+
+    def run_job(self, mesos_offer):
+        """Start a queued Airflow task in Mesos"""
+        offer = mesos_offer.get_offer()
+        tasks = []
+        option = {}
+        offer_cpus = 0
+        offer_mem = 0
+        for resource in offer['resources']:
+            if resource['name'] == "cpus":
+                offer_cpus += resource['scalar']['value']
+            elif resource['name'] == "mem":
+                offer_mem += resource['scalar']['value']
+
+        self.log.debug(
+            "Received offer %s with cpus: %s and mem: %s", offer['id']['value'], offer_cpus, offer_mem
+        )
+
+        remaining_cpus = offer_cpus
+        remaining_mem = offer_mem
+
+        while (
+            (not self.task_queue.empty())
+            and remaining_cpus >= self.task_cpu
+            and remaining_mem >= self.task_mem
+        ):
+
+            key, cmd, executor_config = self.task_queue.get()
+            self.log.debug(executor_config)
+            tid = self.task_counter
+            self.task_counter += 1
+            self.task_key_map[str(tid)] = key
+
+            port_begin = 31000 + tid
+            port_end = 31000 + tid
```

Comment:
What happens if you run more than 35535 tasks? This port number will overflow

Reply:
Port begin and end is just a resource filter for mesos. It will not open a port. It's only to tell mesos that we need a resource with the free port range "port_begin" and "port_end". So, in worst case, mesos will tell the Executor that the port is out of range. But it's a good point. I will remove these filter after I test it.

Reply:
Don't need anymore. :-)